File: src/main/java/com/distributed/sql/utils/Logger.java
package com.distributed.sql.utils;

import org.slf4j.LoggerFactory;

/**
 * Centralized logging utility for the distributed SQL engine
 */
public class Logger {
    private static final org.slf4j.Logger logger = LoggerFactory.getLogger("DistributedSQLEngine");

    public static void info(String message) {
        logger.info(message);
    }

    public static void info(String format, Object... args) {
        logger.info(format, args);
    }

    public static void debug(String message) {
        logger.debug(message);
    }

    public static void debug(String format, Object... args) {
        logger.debug(format, args);
    }

    public static void warn(String message) {
        logger.warn(message);
    }

    public static void warn(String format, Object... args) {
        logger.warn(format, args);
    }

    public static void error(String message) {
        logger.error(message);
    }

    public static void error(String message, Throwable throwable) {
        logger.error(message, throwable);
    }

    public static void error(String format, Object... args) {
        logger.error(format, args);
    }

    public static void trace(String message) {
        logger.trace(message);
    }

    public static void trace(String format, Object... args) {
        logger.trace(format, args);
    }

    // Query lifecycle logging
    public static void queryStart(String queryId, String sql) {
        info("QUERY_START: queryId={}, sql='{}'", queryId, sql);
    }

    public static void queryPlan(String queryId, String plan) {
        info("QUERY_PLAN: queryId={}, plan={}", queryId, plan);
    }

    public static void queryDispatch(String queryId, String workerId) {
        info("QUERY_DISPATCH: queryId={}, workerId={}", queryId, workerId);
    }

    public static void queryComplete(String queryId, long executionTimeMs, int resultRows) {
        info("QUERY_COMPLETE: queryId={}, executionTime={}ms, resultRows={}", 
             queryId, executionTimeMs, resultRows);
    }

    public static void queryError(String queryId, String error) {
        error("QUERY_ERROR: queryId={}, error={}", queryId, error);
    }

    // Worker lifecycle logging
    public static void workerStart(String workerId, String address) {
        info("WORKER_START: workerId={}, address={}", workerId, address);
    }

    public static void workerTaskStart(String workerId, String taskId) {
        info("WORKER_TASK_START: workerId={}, taskId={}", workerId, taskId);
    }

    public static void workerTaskComplete(String workerId, String taskId, long executionTimeMs) {
        info("WORKER_TASK_COMPLETE: workerId={}, taskId={}, executionTime={}ms", 
             workerId, taskId, executionTimeMs);
    }

    public static void workerFailure(String workerId, String reason) {
        warn("WORKER_FAILURE: workerId={}, reason={}", workerId, reason);
    }

    // Fault tolerance logging
    public static void checkpointCreated(String checkpointId, String workerId) {
        info("CHECKPOINT_CREATED: checkpointId={}, workerId={}", checkpointId, workerId);
    }

    public static void checkpointRestored(String checkpointId, String workerId) {
        info("CHECKPOINT_RESTORED: checkpointId={}, workerId={}", checkpointId, workerId);
    }

    public static void retryAttempt(String taskId, String workerId, int attemptNumber) {
        warn("RETRY_ATTEMPT: taskId={}, workerId={}, attempt={}", taskId, workerId, attemptNumber);
    }
}

File: common/src/utils/logger.rs
//! Logging utilities

use tracing::{info, warn, error, debug, trace};

/// Logger utility functions for the distributed SQL engine
pub struct Logger;

impl Logger {
    /// Log query start
    pub fn query_start(query_id: &str, sql: &str) {
        info!("Query started: {} - {}", query_id, sql);
    }

    /// Log query completion
    pub fn query_complete(query_id: &str, execution_time_ms: u64) {
        info!("Query completed: {} - {}ms", query_id, execution_time_ms);
    }

    /// Log query dispatch to worker
    pub fn query_dispatch(query_id: &str, worker_id: &str) {
        debug!("Query dispatched: {} -> {}", query_id, worker_id);
    }

    /// Log worker task start
    pub fn worker_task_start(worker_id: &str, task_id: &str) {
        debug!("Worker task started: {} - {}", worker_id, task_id);
    }

    /// Log worker task completion
    pub fn worker_task_complete(worker_id: &str, task_id: &str, execution_time_ms: u64) {
        debug!("Worker task completed: {} - {} - {}ms", worker_id, task_id, execution_time_ms);
    }

    /// Log shard operation
    pub fn shard_operation(operation: &str, shard_id: &str, details: &str) {
        info!("Shard operation: {} - {} - {}", operation, shard_id, details);
    }

    /// Log system startup
    pub fn system_startup(component: &str, port: u16) {
        info!("System startup: {} on port {}", component, port);
    }

    /// Log system shutdown
    pub fn system_shutdown(component: &str) {
        info!("System shutdown: {}", component);
    }

    /// Log error with context
    pub fn error_with_context(context: &str, error: &str) {
        error!("{}: {}", context, error);
    }

    /// Log warning with context
    pub fn warning_with_context(context: &str, warning: &str) {
        warn!("{}: {}", context, warning);
    }

    /// Log debug information
    pub fn debug_info(message: &str) {
        debug!("{}", message);
    }

    /// Log trace information
    pub fn trace_info(message: &str) {
        trace!("{}", message);
    }
}
